### DISCUSSION TITLE
Support alternative linters such as Biome

### DISCUSSION THREAD

**Discussion by shirakaba** [vercel/next.js#59347] [Ideas]
> [!IMPORTANT]  
> **EDIT:** It took two years, but `next lint` was finally [deprecated](https://nextjs.org/blog/next-15-5#next-lint-deprecation) in Next.js 15.5, and removed in Next.js 16. We did it! 🥳 the discussion below can now be considered resolved.

### Goals

 1. Extract the doctoring logic from `next lint` and move it to its own independent command (e.g. `next doctor`)
2. Allow users to lint using their own linters, such as [Biome](https://biomejs.dev), without wrapping it in a CLI tool like `next lint`

### Non-Goals

1. Providing boilerplate for all possible linters in Next.js itself (instead, it should provide a way for the community and users to provide their own linting boilerplate)

### Background

# The case against eslint and prettier

Eslint is [deprecating formatting rules](https://eslint.org/blog/2023/10/deprecating-formatting-rules/#:~:text=In%20ESLint%20v8.,semicolons%2C%20string%20formats%2C%20etc.) and community momentum for formatting looks set to move [from prettier to Biome](https://prettier.io/blog/2023/11/27/20k-bounty-was-claimed). Thus, the current "format with prettier, lint with eslint" setup (which has always pain-points due to mutually exclusive rules and setting up two different tools) is looking less attractive in the long term than simply "format and lint with Biome"..

# Users can't bring their own linter

[Another discussion](https://github.com/vercel/next.js/discussions/59165) has raised the case for documenting how to integrate Biome, but this isn't merely a matter of documentation, as I'll explain below.

As of today, starter templates (at least [for Turbo](https://turbo.build/repo/docs/getting-started/from-example)) typically configure projects to lint using `next lint`. This prevents users from bringing their own linting setups as linting is dependent on the Next.js CLI.

Unfortunately, we can't simply ditch `next lint`, as it **does more than just linting**. It does some significant (and valuable) doctoring, too.

`next lint`, as of v14.0.4, does the following:
* **it configures some [default values](https://github.com/vercel/next.js/blob/d4d9dc40facff846582de9192f89b5b59c682edf/packages/next/src/cli/next-lint.ts#L20) for the CLI flags to pass to eslint.** Actually, in most cases, it simply defaults to the same default as the eslint CLI, but it _does_ provides React-appropriate default values for the args `--ext` and a Next-appropriate `--cache-location`.
* **it [sanity-checks](https://github.com/vercel/next.js/blob/d4d9dc40facff846582de9192f89b5b59c682edf/packages/next/src/lib/verify-typescript-setup.ts#L36) the repo for you.** Namely, it:
  * checks whether you've installed `typescript`, `@types/react`, and `@types/node`
  * [reconfigures (or creates)](https://github.com/vercel/next.js/blob/canary/packages/next/src/lib/typescript/writeConfigurationDefaults.ts#L118) your `tsconfig.json`:
    * ensures you have the [desired compiler options](https://github.com/vercel/next.js/blob/d4d9dc40facff846582de9192f89b5b59c682edf/packages/next/src/lib/typescript/writeConfigurationDefaults.ts#L21)
    * ensures you have `["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"]` in your `includes` array
    * ensures that you have the Next.js plugin `{ "name": "next" }` in your `plugins` array
    * ensures that you have strict type-checking
    * excludes `node_modules`
  * checks whether `next-env.d.ts` has been generated
  * can optionally run a type-check before any linting
* **it may finally [run eslint](https://github.com/vercel/next.js/blob/d4d9dc40facff846582de9192f89b5b59c682edf/packages/next/src/lib/eslint/runLintCheck.ts#L280).** We can avoid running eslint by deliberately omitting an eslint config and either:
  1. ~~setting `opts.lintDuringBuild: true`~~ (impossible because the value is hard-coded).
  2. setting `opts.lintDuringBuild: false` (which is the default) and selecting `null` when prompted which config to start with next time `next lint` is run. However, this is interactive, so unless we can find a way to pipe the necessary input into it via the shell, I am not sure how we can automate this.

### Proposal

1. Deprecate `next lint` and allow users to use their preferred linter command directly.
2. Move all the doctoring logic into a `next doctor` command that users could optionally run as part of their linting step.

**Comment by JoeSlain** [comment]
Legit request, had to implement biome in my turborepo but still had to let next lint run during linting to keep doctoring functions. Would be nice to be able to chose

**Comment by wottpal** [comment]
This should get more attention. Biome is really becoming a way more performant replacement for the legacy prettier/eslint toolchain.

**Comment by l4b4r4b4b4** [comment]
Is this a WIP or already supported?

**Reply by shirakaba** [reply]
No new issues or PRs containing the word "Biome" have been opened on the Next.js repo since I opened this discussion, so there's no indication yet that this proposal is even under official consideration.

**Comment by davidlbowman** [comment]
Easy +1.

**Comment by admmasters** [comment]
This really needs to happen.

Nextjs used to feel like it was the future, now it seems to be in a bit of a mire:

- Turbopack is taking far too long and other stacks are thriving on Vite (see Remix and almost every other framework)
- Months without members of the core team even responding to valid issues and suggestions like this

Genuinely starting to ask the question is Nextjs really a good choice now?

**Comment by damien-schneider** [comment]
+1

**Comment by eliasdanielr** [comment]
+1

**Comment by John-Dennehy** [comment]
Would love to see more progress on this. Frustrating to see eslint issues during deployment, when it isn't part of my tool chain.

**Reply by Willem-Jaap** [reply]
You can add the `ignoreDuringBuilds` flag to `next.config.js`

```ts
module.exports = {
  eslint: {
    // Warning: This allows production builds to successfully complete even if
    // your project has ESLint errors.
    ignoreDuringBuilds: true,
  },
}
```
https://nextjs.org/docs/app/api-reference/next-config-js/eslint

**Comment by avaltat** [comment]
+1

**Comment by hgrafa** [comment]
+1

**Comment by eliasdanielr** [comment]
+1

**Reply by peterdinis** [reply]
+1

**Comment by HunteRoi** [comment]
+1

**Comment by hanan-yara** [comment]
+1

**Comment by ivanjeremic** [comment]
+1

**Comment by alexmandrikdev** [comment]
+1

**Comment by jakst** [comment]
Please stop commenting with `+1`. It creates a bunch of notification noise for those who subscribe to the issue. Instead, add your support to the discussion by upvoting the top post. (And yes, I'm sorry for creating yet another notification -.-)

**Reply by rajington** [reply]
haha was wrestling with creating another notification myself but had to be said! (also hope this reply doesn't notify everyone!)

depending on how they prioritize their backlog, these `+1`s that cause others to unsubscribe might actually _hurt_ the chances it gets looked at!

really wish GitHub provided more here, I believe they have more fine-grained "only notify me when it's fixed" notifications for issues, but at the same time some discussion conversations offer alternatives or external updates that I'd hate to miss!

**Comment by mcheng-brilliant** [comment]
Hello, we were directed to post to GitHub by Vercel support. We had some questions relevant to this discussion.

We have been using `next lint` as part of our CI to catch linter and formatting issues. However, we discovered that `next lint` was not linting all of our files. For example, `middleware.ts` which, per the docs, lives in the root of our project. We revisited the docs and realized that `next lint` only runs against a fixed set of directories and that list can be manually configured.

Questions we believe that are relevant to this discussion as it pertains to potential design flaws of `next lint` and would be solved by the `next doctor` the split proposed:

1. Why doesn't the default config for `next lint` check all of the known files in a Next.js project? Is it because there's a reason why we shouldn't be linting files like `middleware.ts`? If so, what are those reasons?
    - `next lint` not staying up-to-date with Next.js's own project layout can lead to a false sense of security as you may think all deployed code is linted, but it actually isn't. This wouldn't happen in a standard ESLint configured project.
2. For files that Next.js's docs advise putting into the root directory (or any other directory not in the default `next lint` list), how should we keep our `next.config.js`'s `eslint.dirs` config up-to-date?
    - We attempted to work around the whitelist approach of `next lint` by specifying the root and then attempting to blacklist files/directories using `.eslintignore`. However, that does not appear to work as the ignore file is not used by `next lint`. This also points towards `next lint` being a much less of a full solution than using `eslint` directly.
3. Is there documentation on how we can configure ESLint ourselves with the same flags as `next lint` so that we can use ESLint and take an approach of linting everything except for what's specified in our `.eslintignore` file? This would help us solve question 2 much more easily.

**Reply by Willem-Jaap** [reply]
Hi there, I have struggled with this in the past as well. I'm also very much in favour of a `next doctor` command.

However, I may be able to steer you in the right direction for taking linting in your own hands:

Ignore the default linting behaviour in builds, before builds, run your own:
```tsx
module.exports = {
  eslint: {
    // Warning: This allows production builds to successfully complete even if
    // your project has ESLint errors.
    ignoreDuringBuilds: true,
  },
}
```
https://nextjs.org/docs/app/api-reference/next-config-js/eslint

You can then integrate eslint yourself (e.g. [typescript-eslint](https://typescript-eslint.io/) etc. etc.) in a `eslint.config.js` file (https://eslint.org/docs/latest/use/configure/) 

And run this in your editor or via cli (https://eslint.org/docs/latest/use/command-line-interface)

This custom linting will not include any of the aditional doctoring of next lint.

I really hope this gains traction from maintainers so we are free to use any linting and also have the other powers `next lint` has.

**Comment by warrenbhw** [comment]
+1 on this!

**Comment by iiAbady** [comment]
This is a necessity from now on....

**Comment by simonorzel26** [comment]
+1

**Comment by theklr** [comment]
+1

**Comment by o8e** [comment]
+1

**Comment by Willem-Jaap** [comment]
@leerob answered my question about this on X:
> Definitely want to explore this (Biome looks amazing). It's something that's on our roadmap!

https://x.com/leeerob/status/1820176704662609993

**Reply by wlechowicz** [reply]
Cool, but means next to nothing, except that it's not being worked on. "On the roadmap" could be 2025 or 2055.

**Comment by anagarajah** [comment]
+1 On this to support own lint packages into Next JS

**Comment by stevecastaneda** [comment]
Another +1.  

Just migrated to biome and it's _so much faster_, and didn't realize NextJS was still going to require it's own linting.  😬

**Comment by 0re5ama** [comment]
+1

**Comment by DKeken** [comment]
+

**Comment by abhishekmmgn** [comment]
+1

**Comment by lyleunderwood** [comment]
Just wanted to point out that since Biome does not currently support custom rules, there is not a clear path to feature parity using biome as an alternative to the next eslint plugin. Until biome supports custom rules or the necessary rules are implemented into core biome, switching to biome as a linter in particular would definitely necessitate sacrificing some of the current features of the eslint plugin. This is obviously not necessarily relevant for other linting tools, but particular to Biome.

That being said, disentangling "doctoring" from linting seems like an obvious step forward.

Also please stop +1 posting.

**Reply by admmasters** [reply]
That's a fair comment, to give a slightly more blunt take on the situation, nextjs tying us to eslint feels limiting. To be be honest the days where we can justify migrating config (next is already behind the current version of eslint btw) are over. The eslint 8-9 migration feels extremely self indulgent and breaks a lot of packages, all the projects I work on with biome feel so much cleaner and don't need so many dependencies (we should be optimising for getting the job done as quickly as possible - eslint in 2024 feels like a total Frankenstein of a toolchain)

**Reply by manavm1990** [reply]
> Also please stop +1 posting.

Yeah. I don't get this at all. I mean why people do this stupid stuff. Add some useful input, or just use the emoji reaction, or contribute something! 😮‍💨 Stop cluttering up the 🧵 to where we have to 📜 incessantly to get to anything useful! 👎🏾 

Should be grounds for being blocked.

**Reply by ematipico** [reply]
That's a fair assessment, but it's also fair to let people choose their linter.

While some rules might be different or not implemented, maybe some users don't need them.

Regardless, it's good to give the choice

**Comment by pirmax** [comment]
+1

**Comment by uncomplexity** [comment]
+1 (just annoying all of you)

**Comment by dmythro** [comment]
I'd really love to see some action on this as Biome is much more preferable to me now, and as far as I know to a lot of devs too.

Probably, people behind `@next/eslint-plugin-next` can quickly provide better requirements, and stages for integration.

As far as **Biome already supports some of the rules** from `@next/eslint-plugin-next`, it is possible to setup at least a basic / recommended configuration to extend from as a starting point, to align with ESLint configuration at least partially.

Then, update Next configuration / documentation in order to support Biome as well. Maybe, run Biome instead of ESLint on build if configured etc.

See:
- https://nextjs.org/docs/pages/building-your-application/configuring/eslint#eslint-plugin
- https://biomejs.dev/linter/rules-sources/#nexteslint-plugin-next

**Comment by jxnnyo** [comment]
Is there any update on this? After taking some benchmarks, using Biome in my project would on average save 60x of time linting. Would be nice to know if this is denied/on the roadmap etc?

**Comment by ravi-outcomes-ai** [comment]
+1

**Comment by DarioLopes** [comment]


**Reply by GabenGar** [reply]
As if biomejs won't have to deal with breaking changes in the future.

**Reply by devth** [reply]
Whether a project choses to break their users with backward incompat changes is a decision. I think eslint lost trust from the community through years of unnecessary backward incompat changes. The hope is that biome would provide better stewardship and carefully consider impact to users.

**Reply by GabenGar** [reply]
You seem to be accustomed to a commercial view of semver major being just an excuse for the blog post and/or a conference when in reality it is used exactly for the purpose of signalling breaking changes to consumers in just a single string. It's pretty bizzare to expect a tool, which existed before ESM even becoming a spec, to retain the legacy behaviour across 9 major versions.
Biome itself [doesn't provide a backward compat guarantee](https://biomejs.dev/internals/versioning) even between patch versions, so I guess don't use linters at all.

**Reply by devth** [reply]
Good point, but I'll give Biome a chance to slowly burn me over a decade+ first.

**Comment by Willem-Jaap** [comment]
Hi all,

You can implement your own linter (biome) just fine. There is nothing in Next.js which is blocking you from using Biome.

Follow these steps:

- [ ] Remove eslint and any related dependencies (probably prettier too)
- [ ] Add the following to your next.config

```tsx
  eslint: {
    // Warning: This allows production builds to successfully complete even if
    // your project has ESLint errors.
    ignoreDuringBuilds: true,
  },
```
https://nextjs.org/docs/app/api-reference/next-config-js/eslint

- [ ] Install biome: https://biomejs.dev/guides/getting-started/ 
- [ ] Add the biome script to your package.json
```json
{
  "scripts": {
    "lint": "biome check ."
  }
}
```

- [ ] Configure biome to your liking.
- [ ] Run the biome script as your linting step and before build
- [ ] 🎉  Done

I have seen a lot of people talking about a mismatch of linting in builds in other issues. This is not related to eslint, rather the typescript compiler which runs during `next build`. If you're setting up a linting step in a CI/CD pipeline you probably want to add a 'type-check' step before build which runs the typescript compiler to check for typing issues: 
```json
{
  "scripts": {
     "type-check": "tsc --noEmit"
  }
}
```

> [!WARNING]  
> When implementing this yourself you are missing out on the doctoring logic of `next lint` (see issue description) but in my experience this is fine in most cases

I still really would love first party biome support!

**Reply by shirakaba** [reply]
> but in my experience this is fine in most cases

Except all the cases that `next lint` handles, yes

**Reply by Willem-Jaap** [reply]
I know that, I'm also in favor of this being resolved correctly. Apart from that doctoring logic there is nothing that is stopping you from using biome in Next.js as a linter and formatter which is what people want, it seems.

**Reply by lyleunderwood** [reply]
As somebody who is already using biome with next and currently doesn't have doctoring, I'm subscribed to this discussion because this discussion is about separating out doctoring from linting.

**Reply by wlechowicz** [reply]
Nice write-up, but completely missed the point.
> There is nothing in Next.js which is blocking you from using Biome.

but
> When implementing this yourself you are missing out on the doctoring logic of next lint

meanwhile this is the whole point of this entire discussion.

Nobody said "you can't use Biome".

**Reply by lyleunderwood** [reply]
> Nobody said "you can't use Biome".

Maybe it's the title on the discussion:

> Support alternative linters such as Biome

Depending on your definition of "support," next does indeed support alternative linters such as Biome. The very first thing said in the description, however, is:

> Extract the doctoring logic from next lint and move it to its own independent command (e.g. `next doctor`)

A more appropriate title might be,

> Improve DX when using alternative linters such as Biome

There are really two parts to this, as I see it:

1. Separating out doctoring, and
2. Some minimal documentation about bringing your own linter.

**Reply by bitttttten** [reply]
what's the difference between `next lint` and what biome can do? what are we missing from `next lint` that is not in biome already?

**Reply by lyleunderwood** [reply]
See the list of doctoring operations listed in the OP.

**Comment by felipedeboni** [comment]
Looking forward to this too as I am sick of ESLint breaking changes and wasting a lot of time on something that should be stable enough.

Apart from that Biome is way faster then Eslint and Prettier.

**Comment by EfstathiadisD** [comment]
I am not sure what is stopping you. Are you using next lint? Cause that's being terrible from the start.

I have my own fmt command that uses biome. You can do it. I guess y want an option on next lint to not do the work hourselves. Right?

**Reply by manavm1990** [reply]
Well, we're using Next for everything else, so it makes sense to have this option taken care of also.

And, yeah, configuration 🔧  details like that aren't the best use of time.

**Comment by alicercedigital** [comment]
+1

Biome is like a car, while ESLint is a horse. Cars are a recent invention and may not yet be the perfect replacement for horses, but the shift is inevitable. Eventually, we'll have to bid farewell to the old horse.

**Comment by Zerebokep** [comment]
Agree, I've switched completely to biome after eslint's v9 f*ckup, I'll never touch eslint again. Biome's DX is awesome.

**Reply by felipedeboni** [reply]
By any chance are you using it on a large monorepo? How bad was the migration time wise?

**Reply by dmythro** [reply]
> By any chance are you using it on a large monorepo? How bad was the migration time wise?

I did a migration for a large monorepo once, it had a lot of changes but in general all positive. When it ignored all eslint specific comments in code, I had a chance to reconsider some code and in many places actually made it better. After migration, lint/format is so much faster so it was 99% worth it. Not 100% because some checks are probably missing, not sure they are benefitial though in that specific case.

Configuration/infra migration was super easy though. But code wise, I was trying to do it right and invested some time.

**Reply by Willem-Jaap** [reply]
Same here, migrated some pretty large monorepo's (500+ files) to Biome. Changes in almost every file. Tried to keep the formatting config as close to prettier as possible. Errors that were reported were generally quite useful, even catched some bugs.

Changes that are required are mostly not too difficult, added some biome ignores to look at later

Am missing some eslint features but it was definitely worth it, I'd estimate we save 1000 minutes on CI time a month by switching to Biome

**Reply by devth** [reply]
The most insane part about biome is defaulting to tabs (LOL) but otherwise it's incredible.

**Reply by Zerebokep** [reply]
I did 5 migrations so far, biome's learning curve is very flat, everything works out of the box, prettier is also not needed anymore. At your first migration there might be some rules missing, but biome had a replacement for anything I use. The majority of work is to replace eslint comments, plus code optimizations here and there which eslint didn't catch (e.g. foreach vs for..in).

Performance is also WAY better than eslint, I guarantee you that you will not walk back to eslint after you tried it.

**Reply by devth** [reply]
Did you need to manually enable a bunch of rules to make it work well with React, like useEffect dependency arrays, or anything else? I'm slightly worried that when I switch I'll have less rules being checked but won't really know they're missing until I discover a bug down the road.

**Reply by Zerebokep** [reply]
I only have 7 rules added in our react project where 3 of them disable some of the "too strict" default rules of biome, useEffect dependencies are checked by default.

You might check the biome docs, all (default) rules are listed there. I've also seen some best practices biome configs in the wild.

**Comment by felipedeboni** [comment]
I am almost finishing the migration to Biome while keeping `next lint`. It's a bit hacky, but it's what it's.

All that needs to be done is to patch `eslint-config-next` and remove everything you don't care about.

This is how my patch looks like:

```patch
diff --git a/index.js b/index.js
index 4c95a88ad4fba7c4257ab5071dbc581d63ee55b4..d5cf85a4aa52dc46d28adbf38ca7b5e44ca6d1d0 100644
--- a/index.js
+++ b/index.js
@@ -31,9 +31,7 @@ sortedPaths.push(...keptPaths)
 const hookPropertyMap = new Map(
   [
     '@typescript-eslint/eslint-plugin',
-    'eslint-plugin-import',
-    'eslint-plugin-react',
-    'eslint-plugin-jsx-a11y',
+    'eslint-plugin-import'
   ].map((request) => [
     request,
     require.resolve(request, { paths: sortedPaths }),
@@ -54,30 +52,9 @@ require('@rushstack/eslint-patch/modern-module-resolution')
 
 module.exports = {
   extends: [
-    'plugin:react/recommended',
-    'plugin:react-hooks/recommended',
     'plugin:@next/next/recommended',
   ],
-  plugins: ['import', 'react', 'jsx-a11y'],
-  rules: {
-    'import/no-anonymous-default-export': 'warn',
-    'react/no-unknown-property': 'off',
-    'react/react-in-jsx-scope': 'off',
-    'react/prop-types': 'off',
-    'jsx-a11y/alt-text': [
-      'warn',
-      {
-        elements: ['img'],
-        img: ['Image'],
-      },
-    ],
-    'jsx-a11y/aria-props': 'warn',
-    'jsx-a11y/aria-proptypes': 'warn',
-    'jsx-a11y/aria-unsupported-elements': 'warn',
-    'jsx-a11y/role-has-required-aria-props': 'warn',
-    'jsx-a11y/role-supports-aria-props': 'warn',
-    'react/jsx-no-target-blank': 'off',
-  },
+  plugins: ['import'],
   parser: './parser.js',
   parserOptions: {
     requireConfigFile: false,

```

Also you lint script should be updated to: 

```
"lint": "biome lint . && next lint"
```

Output becomes:

```
➜ pnpm lint

> lint /path/to/project
> biome lint . && next lint

Checked 134 files in 8ms. No fixes applied.

./path-to-file.tsx
24:5  Warning: Using `<img>` could result in slower LCP and higher bandwidth. Consider using `<Image />` from `next/image` or a custom image loader to automatically optimize images. This may incur additional usage or cost from your provider. See: https://nextjs.org/docs/messages/no-img-element  @next/next/no-img-element

info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/app/api-reference/config/eslint#disabling-rules
```

Only downside to that is the editor don't show up any ESLint violations, only through command line.

---

Additionally, you can make your team life easier with the following additions to the project (if using vscode).

project `settings.json`:
```json
{
  "biome.enabled": true,
  "eslint.enable": true,
  "sort-imports.on-save": false,
  "editor.defaultFormatter": "biomejs.biome",
  "editor.formatOnPaste": false,
  "editor.formatOnSave": true,
  "editor.codeActionsOnSave": {
    "quickfix.biome": "always",
    "source.organizeImports.biome": "always"
  },
  "[typescriptreact]": {
    "editor.defaultFormatter": "biomejs.biome"
  },
  "[typescript]": {
    "editor.defaultFormatter": "biomejs.biome"
  },
  "[json]": {
    "editor.defaultFormatter": "biomejs.biome"
  },
  "[css]": {
    "editor.defaultFormatter": "biomejs.biome"
  }
}

```

project: `extensions.json`

```json
{
  "recommendations": [
    "biomejs.biome"
  ]
}
```

If you have many projects using ESLint and Prettier and will migrate one at time and is annoyed by the error message that shows up when biome isn't detected, take a look [here](https://github.com/biomejs/biome-vscode/discussions/492)

**Reply by PaulRBerg** [reply]
Silly question but how do you patch `eslint-config-next` like you did? Is there a fork?

**Reply by felipedeboni** [reply]
@PaulRBerg there are a couple options depending on your package manager.

With pnpm or yarn just execute one of the following and follow the instructions.
```
pnpm patch eslint-config-next
yarn patch eslint-config-next
```

With yarn 1 or npm you need to install [patch-package](https://www.npmjs.com/package/patch-package), if that's the case just follow patch-package guide.

**Comment by lveillard** [comment]
Following so I can throw a big party once we get rid of ESLint. ESLint 8 was painful enough, but 9.x has exceeded all my worst expectations.

**Reply by devth** [reply]
+1000. Delete eslint.

**Comment by Netail** [comment]
Biome is working hard on version 2.0, with plugin support. Which the NextJS team is perhaps waiting on?

**Reply by suiramdev** [reply]
This seems to be the case, as they have just deprecated the `vercel/style-guide` repo.
I read somewhere on X, from Lee Robinson, that he recommended using Biome instead.

Oh and they are also sponsors of Biome

**Comment by destroyer22719** [comment]
I think that we might be getting nextjs with biome? Because first the biome 2.0 allows custom plugins, which is a huge step in the right direction, but the biggest one of all is domains for nextjs: https://next.biomejs.dev/linter/domains/#next

They're still missing some features from the original eslint plugin: https://nextjs.org/docs/app/api-reference/config/eslint

But given that Vercel is sponsoring Biome and that they're using it in a monorepo, could this be a sign that biome will soon be an option for nextjs when biome v2.0 is out of beta?

**Comment by TeoChi-CF** [comment]
Another option, similar to https://github.com/vercel/next.js/discussions/59347#discussioncomment-12154120 is to revert to base eslint config file
```
{
  "extends": "next"
}
```
and if after applying the required Biome linter changes there are some differences/conflicts that come up during next build, just disable those specific eslint rules:
```
{
  "extends": "next",
  "rules": {
    "react-hooks/exhaustive-deps": "off"
  }
}
```
No patching or any other hackish solution required. You'll have to keep the eslint config file till a better solution is implemented though.

**Reply by chriscoomber** [reply]
Going further, your `.eslintrc.json` can just be:

```json
{
  "ignorePatterns": ["/*"]
}
```

if you want it to truly do nothing.

**Comment by destroyer22719** [comment]
Hey guys, so Biome v2 is finally stable and that means you can now make custom plugins, does Vercel now have plans to give users the option to use biome for `next lint` and `next build`?

https://biomejs.dev/blog/biome-v2/

**Reply by dmythro** [reply]
> Hey guys, so Biome v2 is finally stable and that means you can now make custom plugins, does Vercel now have plans to give users the option to use biome for `next lint` and `next build`?
> 
> 
> 
> https://biomejs.dev/blog/biome-v2/
> 
> 

As far as I've seen, Vercel is a prestige sponsor of Biome now, so 🤞

**Reply by abed-daloopa** [reply]
I think there's no need for them to create a plugin.

Next is a first-class "domain" in Biome v2

https://biomejs.dev/linter/domains/#next

**Comment by Yohannfra** [comment]
Hi, I know this discussion is mainly about biome but last week there is also [Oxlint](https://oxc.rs/docs/guide/usage/linter.html) that came out in [1.0.0](https://oxc.rs/blog/2025-06-10-oxlint-stable.html) and it works really well for my turborepo and nextjs codebase.
There is even an official nextjs plugin (https://oxc.rs/docs/guide/usage/linter/plugins.html).

**Reply by dmythro** [reply]
I've heard about oxlint but didn't try yet — and no, I think this thread is about any tool people like, and to be able to use it seamlessly with Next :) 

The more you mention the tool you like — the more chances for a better integration.

**Reply by Yohannfra** [reply]
You should give it a try with ```npx oxlint@latest``` then, it's *really* fast.

Btw this is my **.oxlintrc.json** if you want to try with more rules

```json
{
  "plugins": ["typescript", "unicorn", "oxc", "import", "node", "promise", "react", "react-perf", "nextjs"],
  "rules": {
    "no-duplicate-imports": "error",
    "typescript/no-explicit-any": "error",
    "eqeqeq": "error",
    "curly": "error",
    "typescript/no-unused-vars": "warn",
    "typescript/explicit-function-return-type": "error",
    "typescript/no-empty-interface": "off",
    "typescript/consistent-type-imports": "error",
    "typescript/require-await": "warn",
    "react/self-closing-comp": "error",
    "react/exhaustive-deps": "error",
    "react/jsx-no-useless-fragment": "error",
    "nextjs/no-img-element": "off",
    "no-console": "off",
    "no-alert": "warn",
    "no-else-return": "error"
  }
}
```

A simple benchmark of the two commands on my nextjs codebase:

```bash
# oxlint
pnpm run oxlint  0.92s user 0.60s system 249% cpu 0.612 total

# eslint
pnpm run lint  30.97s user 2.79s system 164% cpu 20.583 total
```

Less than a second for oxlint and 30+ seconds for eslint

**Comment by n2k3** [comment]
It happened: https://nextjs.org/blog/next-15-5#next-lint-deprecation

**Reply by aybangueco** [reply]
![FAgwmvtVIAI7OaV](https://github.com/user-attachments/assets/67c3b878-df1b-4b4c-bf3c-d9c0c0622da3)

**Comment by bgw** [comment]
As discussed, `next lint` was removed in Next 16, so I'm closing this discussion as resolved.
https://nextjs.org/blog/next-16#breaking-changes-and-other-updates

**Reply by nikelborm** [reply]
Will `next doctor` replace it, as it's written in the proposal?
